### ISSUE TITLE
Configuration validation backed by schema

### ISSUE THREAD

**Issue by medikoo** [serverless/serverless#6562]
I've just put `websockets` instead of `websocket` and deploy was successful, altough no Websocket API and  no binding  was setup.

It'll be great if such issues are picked and communicated by the framework, Still it should be done with respect to plugins ,which may support events that are not supported at Framework Core.

--- 

Ideally we should have fine grain validation towards `serverless.yml` (backed with library as [@hapi/joi](https://hapi.dev/family/joi/)), which can be extended by the plugins.

With v1 we can introduce a non-breaking version that'll just show warnings, while with next major it can throw meaningful errors.

Validation should be made after all variables are resolved.

---

As introducing complete validation for whole config as we support now seems non-trivial and very time consuming task. It'll be good to start with something minimal that provides basic validation and just allows additional properties in places where schema is not complete.

It'll allow us to deliver early, and then improve and fine tune schema on the go, and cover new functionalities with complete schema upfront.

--- 

#### Where validation logic should be attached?

Validation should not be configured as an internal plugin but be a part of a core (so be incorporated into engine which is also responsible for load of config, resolving variables, initializing plugins). It should be considered as a core feature.

#### When validation should happen?

Ideally validation should happen after all plugins were initialized and variables resolved but before life-cycle engine jump in.

Technically it's when current `service.validate()` happens: https://github.com/serverless/serverless/blob/7abb23edc8dfbe5005ac716aa137330741759929/lib/classes/Service.js#L174-L202

Which already tries to do some validation which should we replace and back with new schema validator

#### What objects should validation cover?

Some problem is that current config handling is not clean. We do not have one object with which we work, but we span config parts on our service instance here: https://github.com/serverless/serverless/blob/7abb23edc8dfbe5005ac716aa137330741759929/lib/classes/Service.js#L98-L131

 and then work with that. Variables resolution is made on `service` instance.

Ideally after Variables resolution is made. I think we should prepare a plain object that derives from service as follows:

```javascript
const configToValidate = {
  service: service.serviceObject,
  app: service.app,
  org: service.org,
  custom: service.custom,
  plugins: service.plugins,
  resources: service.resources,
  functions: service.functions.
  provider: service.provider,
  package: service.package,
  layers: service.layers,
  outputs: service.outputs
}
```

And validate that.

#### What validation library should we use?

In consideration:

- [@hapi/joi](https://hapi.dev/family/joi/)
- [ajv](https://ajv.js.org/)

We need a validation tool that allows to easily in generic way extend base schema (e.g. define a new property on some object which is a deep nested property in global schema).

So we don't have to upfront define those eventual extension points, as otherwise we can easily run into maintenance issues in a future (it's likely that we will need more and more customization options on the way)

I'm not that familiar with`@hapi/joi` and I'm not sure if it provides easy means for that (by first look, it appears to me it doesn't, but I might have missed something)

We may also consider: https://github.com/epoberezkin/ajv which is based on JSON schema standard (so schema on it's own is a JSON, therefore extensible). It also supports `$merge` instructions dedicated for that https://github.com/epoberezkin/ajv#merge-and-patch-keywords

Before we jump into implementation, let's prepare a working a example, where we have some example schema with deep nested objects, and then we extend some deep objects with schema for additional properties.

Having that we could be more confident that we've chosen right schema validator.

**Comment by keithalpichi** [comment]
I agree with this. I've had a similar experience recently where I was configuring the deployment package using the `exclude` key.

It was:
```
someLambda:
  handler: src/handler.someLambda
  exclude:
      - 'some-file.txt'
```
The library packaged it up correctly but did not recognize my mistake. As a result, the `some-file.txt` file was not excluded.

Of course it should have been this (with the parent key `package`):
```
someLambda:
  handler: src/handler.someLambda
  package:
    exclude:
        - 'some-file.txt'
```

I didn't realize this mistake until months later when more files that were added to this list was suppose to be excluded. So each file/directory added to the size of the deployment package. Builds started to fail due to `too many open files` errors.

Has there been discussion around validation of the configuration files? Maybe provide warning to stdout or fail completely.

**Comment by medikoo** [comment]
@keithalpichi yes, we've discussed that internally. It's highly wanted it, as it'll pick tons of configuration issues we observe.

Still, it's a time-taking task, and our internal resources are limited, so we're not sure when we'd be able to provide that, anyway PR is highly welcome!

_I've updated issue description to make this issue in general about configuration validation, and not about specific case_

**Comment by preshetin** [comment]
I am working on this. Hopefully, there will be a PR soon.

For the validation at the 1-st stage, before variable resolution, I found only one case in https://github.com/serverless/serverless/issues/7203 So not sure if schema validation is necessary at this stage.

For the 2-nd stage, after variables resolution, there would be schema validation backed by  @hapi/joi

Hoping the resulting PR would be good enough.

**Comment by medikoo** [comment]
Thanks @preshetin for giving that a spin! It's an ambitious task, but I think following with small steps, we can come up in reasonable time with something valuable.

> @medikoo if you think that 1-st stage validation (before variable resolution) should also be backed by schema also, please let me know.

I think we can skip that (assuming that 1-st stage validation is purely about YAML conformity, which is done by YAML parser). Problem described in #7203 will be way better solved with #7207.

I've updated the main description to just include: _Validation should be made after all variables are resolved._

**Comment by preshetin** [comment]
@medikoo no problem, I've also removed the small part about non-schema validation in https://github.com/serverless/serverless/pull/7335/commits/5997cb4f6c878c300ffbde5585e62acfd7faf9e6.

I will continue the discussion about this issue in PR thread

**Comment by medikoo** [comment]
I've updated main description with basic implementation proposal.

It'll be great to decide now on right schema validation tool (see expectations in main description), if anyone has experience with any schema tools, any feedback is highly welcome.

**Comment by jthomerson** [comment]
I've had really good success using [AJV](https://www.npmjs.com/package/ajv) for schema validation in multiple production APIs, so I'm a big fan of it.

A couple other comments about the implementation:

1. Plugins will need to be able to extend the validation because they can add custom events, or configuration key requirements.
2. Validation should take place only after plugins have at least initialized. I have plugins that modify the configuration that's read from serverless.yml. For example, one "boilerplate" plugin takes the config that was read from the file and augments it so that we don't have to copy/paste a ton of the boilerplate config (custom variables, defaults for configs, etc) to all of our 60+ services.

**Comment by medikoo** [comment]
@jthomerson thanks for comment. I'm also leaning towards AJV now.

Plan is to implement it in a way all constraints you mention are met: Plugins will be able to extend the schema. Validation will happen after all plugins will initialize (after variables resolution, but before life-cycle engine starts)

**Comment by preshetin** [comment]
Thanks for the feedback.

Let me share what I've come so far. In summary, for deep nested schemas I was unable to perform merging. But I've got a working example when the Framework allows only predefined ways for plugins to extend a schema.

## Merging deep nested schemas

I was unable to provide a working example of merging deeply nested schemas. Both in Joi and in Ajv. 

For Joi, I believe this is not possible. I stopped attempts after running into https://github.com/hapijs/joi/issues/978#issuecomment-244039949. 

For Ajv, I got stuck at the very start. I was unable even to merge a sequence of schemas. Provided there are two plugins that extend the schema:

```javascript
var Ajv = require('ajv');

var awsEventSchema = {
    "$id": "awsEventSchema#",
    "additionalProperties": false,
    "properties": {
      "http": { "type": "string" },
      "websocket": { "type": "string" },
    }
  };
  
  var extendedByFirstPlugin = {
    "$id": "extendedByFirstPlugin#",
    "$merge": {
      "source": { "$ref": "awsEventSchema#" },
      "with": {
        "properties": {
          "newEvent": { "type": "string" }
        }
      }
    }
  };
  
  var extendedBySecondPlugin = {
    "$id": "extendedBySecondPlugin#",
    "$merge": {
      "source": { "$ref": "extendedByFirstPlugin#" },
      "with": {
        "properties": {
          "anotherEvent": { "type": "string" }
        }    
      }      
    }        
  };         
  
  var validData = {
    http: 'get some-path',
    newEvent: 'provided by first plugin',
    anotherEvent: 'another'
  };     

var ajv = new Ajv({schemas: [awsEventSchema, extendedByFirstPlugin, extendedBySecondPlugin]});
require('ajv-merge-patch')(ajv);
         
const validate = ajv.getSchema('extendedBySecondPlugin#');
const result = validate(validData)
  
console.log('result:', result, validate.errors);
/*
result: false [
  {
    keyword: 'additionalProperties',
    dataPath: '',
    schemaPath: '#/additionalProperties',
    params: { additionalProperty: 'anotherEvent' },
    message: 'should NOT have additional properties'
  },
  {    keyword: '$merge', ...},
  {    keyword: '$merge', ...}
]
*/
```
In the above example, the validation of `validData` is negative. When there is only one extension, there result is fine. But for two merges I was unable to make things work.

Though I am not sure how to use http pointers (also, I'm not sure what `#` means id schema id's).

## Extending a schema with a list of predefined ways

I was more lucky with crafting an example when the Framework defines strict ways to extend a schema. For example, these ways of schema extension can be created: 
- a plugin can add a schema for its parameters in `custom` section (I suppose, this covers the majority of use cases)
- a plugin can add schema for its new events 
- a plugin can add schema for a property to a specific event, etc.

The example below uses Ajv lib. This also can be done with Joi lib. This way, a plugin just passes piece of validation schema to a new `SchemaModifier` class:

```javascript
 class NewEventPlugin {
    constructor(serverless) {           
      const newEventSchema = {                                                            
        type: 'object',                                                                   
        properties: {                                                                     
          someProp: { type: 'sting' },
          anotherProp: {type: 'number' }                                                  
        }                                                                                 
      };

      serverless.schemaModifier.addEventSchema('newEvent', newEventSchema);          
    }                                                                                     
  }
```

The `SchemaModifier` is a new class that lives in `libs/classes/SchemaModifier.js` and looks like this:

```javascript
  class SchemaModifier {                                                                  
    constructor() {                                                                       
      this.awsEventPieces = [];                                                           
      this.customPropertyPieces = [];
      // ...                                                                              
      // list of other ways to modify schema                                              
    }                                                                                     
                                                                                          
    addEventSchema(name, schema) {                                                        
      this.awsEventPieces.push({name, schema});                                           
    }       

   addCustomPropertySchema(name, schema) {                                                        
      this.customPropertyPieces.push({name, schema});                                           
   }       
 }                  
```
The purpose of this class is to store pieces of schemas added by plugins. After all plugins have added their pieces of schemas, the schema json object is assembled in `validate()` method of `Service` class:

```javascript
       let awsEventSchema = {
         type: 'object',
         parameters: {
           http: { type: 'string' },
           websocket: { type: 'object', parameters: {} },
           //...
           //list of allowed events goes here
         }
       } 
  
      const awsEventPieces = this.serverless.schemaModifier.awsEventPieces;
  
      for (let piece of awsEventPieces) {
        awsEventSchema.parameters[piece.name] =  piece.schema
      }
```
----

Please let me know what you think about it

**Comment by medikoo** [comment]
Great thanks @preshetin for checking this.

I was working once with AJV, and I think we can achieve it with it. Still I wouldn't rely on `$merge` operations, at least it appears as a bit too complicated, when on JS grounds we can simply extend JSON schema by adding new items in places.

I've prepared some bare example which introduces basic schema for `custom` and `functions` objects:

```javascript

const functionNamePattern = "^[a-zA-Z0-9-_]+$";

const schema = {
  type: "object",
  properties: {
    custom: {
      type: "object",
      properties: {}
    },
    functions: {
      type: "object",
      patternProperties: {
        [functionNamePattern]: {
          type: "object",
          properties: {
            handler: { type: "string" },
            events: {
              type: "array",
              items: {
                anyOf: [
                  {
                    type: "object",
                    properties: {
                      httpApi: {
                        anyOf: [
                          { type: "string" },
                          {
                            type: "object",
                            properties: {
                              method: { type: "string" },
                              path: { type: "string" }
                            },
                            required: ["path"],
                            additionalProperties: false
                          }
                        ]
                      }
                    },
                    required: ["httpApi"],
                    additionalProperties: false
                  }
                ]
              }
            }
          },
          required: ["handler"],
          additionalProperties: false
        }
      },
      additionalProperties: false
    }
  },
  additionalProperties: false
};
```

To make extensions easier aside of exposing global schema, we can expose some helpers for plugins to extend:

```javascript
const subSchemas = {
  custom: schema.properties.custom.properties,
  functionEvents:
    schema.properties.functions.patternProperties[functionNamePattern]
      .properties.events.items.anyOf
};

const defineCustomProperty = (name, schema) => {
  subSchemas.custom[(name, schema)];
};

const defineFunctionEvent = (name, schema) => {
  subSchemas.functionEvents.push({
    type: "object",
    properties: {
      [name]: schema,
      required: [name],
      additionalProperties: false
    }
  });
};

```

Having that it should be straightforward for plugin authors to add their schemas as follows:

```javascript
defineCustomProperty("pluginNamespace", {
	type: "object",
	properties: {},
	additionalProperties: false
})
defineFunctionEvent("pluginEvent", {
	type: "object",
	properties: {},
	additionalProperties: false
})
```

To ensure plugins will not break anything (by accident), we may also freeze some schema parts, so they cannot be updated, e.g.:

```javascript
Object.freeze(schema);
```

Having schema as above, we can validate config as:

```javascript

const ajv = new Ajv();
const validate = ajv.compile(schema);
validate(userConfig);
if (validate.errors) console.log("Config is invalid!", validate.errors);
```

What will be challenging though is displaying those errors in UX friendly way.
As I observe locally we will have some quirks to workaround.

What do you think?

**Comment by preshetin** [comment]
Thanks, @medikoo for sharing these useful ideas.

I am feeling we are much closer to agreeing with the implementation points. 

The approach to use AJV without $merge feature seems much easier. Also, as the schema is a plain JSON object that plugins would have access to, there will be a possibility for them to extend it without any predefined ways by the Framework. However, to improve plugin author's DX, it'd be nice to have the mentioned helpers. 

For those helpers, I am also thinking it'd be great to add default schemas so that the author may just define for example a new event name without specifying its schema shape. The default schema could be `anyOf` object or string.

Freezing parts of the schema also seem reasonable. I am thinking it'd be good to freeze all parts of schema except `custom` and `functions` parts. Though I may miss something as I personally haven't used plugins much.

Here are two more points it'd be better to agree.

#### How would the schema be structured?

The above `schema` code snippet seems a good foundation for the schema object implementation. I assume it will be stored in a `schema.js` file rather than `schema.json`. This way, the schema could be split into several files (for readability reasons) and imported with the `require` method. Having that said, I also assume AJV's $ref feature for file splitting will not be used. Otherwise, we would need to expose all `.json` files to plugin authors which seems overwhelming to me. 

#### How plugins would access the schema and helper methods?

As the plugin has access to the `serverless` instance, the schema has to be somewhere inside it. I suggest we create a new class that would contain schema instance as well as helper methods.

Another good place is to add schema property and helper methods to the `Service` class. The size of the service class is about 300 lines which seems not so much.

----

The bottom line is that I feel confident to implement the proposed way. Though if there are any doubts feel free to express them.

**Comment by medikoo** [comment]
> I am also thinking it'd be great to add default schemas so that the author may just define for example a new event name without specifying its schema shape. The default schema could be anyOf object or string.

In this case I would rather expect author of plugin to provide relaxed schema specifically. I believe we should rather encourage plugin authors to provide complete schemas, as without that this feature will be ineffective.
We may eventually provide an example of relaxed schema (as a starting point template) in documentation.

>  am thinking it'd be good to freeze all parts of schema except custom and functions parts

Yes, it's probably a good idea.

> The above schema code snippet seems a good foundation for the schema object implementation. I assume it will be stored in a schema.js file rather than schema.json

Yes, I agree, I propose to put it into `lib/configSchema/index.js` (where `index.js` is access point for complete schema, but folder will have schema distributed into many files, for convenience of maintenance)

> I also assume AJV's $ref feature for file splitting will not be used

Yes, as our approach is to pre-prepare complete schema, before we pass it to AJV, I think it's easier to not rely on $ref.

Although we may at some point decide to rely on `$ref`, where e.g. we approach a need of creating custom types, to be reused. Still I would definitely not use `$ref` for parts not intended for multiple reuse.

> I suggest we create a new class that would contain schema instance as well as helper methods.

Good idea, I would probably name it `ConfigSchemaHandler` (any better ideas? I think `ConfigSchema` is not great, as it suggests plain schema).
And then it should be placed in `lib/classes/ConfigSchemaHandler.js`

Then instance should be exposed at `serverless.configSchemaHandler`, and `service.validate` could internally just run `serverless.configSchemaHandler.validateConfig(config)`

**Comment by DaveLo** [comment]
EventBridge events can't be generated from imports(#7063 ):
```yaml
functions:
  testFunc:
    handler: test-func.handler
    events:
      - eventBridge:
          eventBus:
            Fn::ImportValue: TestBusName
```
This curently throws an error which makes the cause unclear because the code is evaluating as a string:

https://github.com/serverless/serverless/blob/103fdacc294ab87f4bd079847d05d9448fd4b494/lib/plugins/aws/package/compile/events/eventBridge/index.js#L133-L135

I threw together a PR for adding error handling (#7367) and @medikoo asked me to instead comment about it so it could be covered as part of this validation project.

**Comment by preshetin** [comment]
>  I would probably name it ConfigSchemaHandler (any better ideas? I think ConfigSchema is not great, as it suggests plain schema).

I think this is a good name.

> What will be challenging though is displaying those errors in UX friendly way.
As I observe locally we will have some quirks to workaround.

Yes, I also noticed that the error message string should be composed of the error objects. For, example, in case of invalid `websocketssss` event name it generates too many error objects:
```javascript
[
  {
    keyword: 'additionalProperties',
    dataPath: ".functions['someFunc'].events[2]",
    schemaPath: '#/properties/functions/patternProperties/%5E%5Ba-zA-Z0-9-_%5D%2B%24/properties/events/items/anyOf/0/additionalProperties',
    params: { additionalProperty: 'websocketssss' },
    message: 'should NOT have additional properties'
  },
  {
    keyword: 'additionalProperties',
    dataPath: ".functions['someFunc'].events[2]",
    schemaPath: '#/properties/functions/patternProperties/%5E%5Ba-zA-Z0-9-_%5D%2B%24/properties/events/items/anyOf/1/additionalProperties',
    params: { additionalProperty: 'websocketssss' },
    message: 'should NOT have additional properties'
  },
  {
    keyword: 'additionalProperties',
    dataPath: ".functions['someFunc'].events[2]",
    schemaPath: '#/properties/functions/patternProperties/%5E%5Ba-zA-Z0-9-_%5D%2B%24/properties/events/items/anyOf/2/additionalProperties',
    params: { additionalProperty: 'websocketssss' },
    message: 'should NOT have additional properties'
  },
  {
    keyword: 'anyOf',
    dataPath: ".functions['someFunc'].events[2]",
    schemaPath: '#/properties/functions/patternProperties/%5E%5Ba-zA-Z0-9-_%5D%2B%24/properties/events/items/anyOf',
    params: {},
    message: 'should match some schema in anyOf'
  }
]
```
The solution I'm thinking should only display one message with text like this:
```
"websocketssss" should NOT have additional properties (path: ".functions['someFunc'].events[2]")
``` 
This can be achieved by `ConfigSchemaHandler`'s `composeErrorMessages` method like this:
```javascript
  /*
   * For error object structure, see https://github.com/epoberezkin/ajv#error-objects
   */
  composeErrorMessages(errors) {
    let result = [];
    for (const error of errors) {
      let message = '';
      const defaultMessage = `Error: ${error.message} (path: ${error.dataPath})`;
      switch(error.keyword) {
        case 'additionalProperties':
          message = `"${error.params.additionalProperty}" ${error.message} (path: ${error.dataPath})`;
          break;
        case 'anyOf':
          if (errors.filter(err => err.keyword !== 'anyOf').length) {
            break;
          } else {
            message = defaultMessage;
            break;
          }
        // todo: add other cases as we run into them
        default:
          message = defaultMessage
          break;
      }
      result.push(message);
    }

    return [...new Set(result.filter(v => v !== ''))];
  }

```
In case we find that such approach creates poor messages, we can fix such cases with the help or [epoberezkin/ajv-errors](https://github.com/epoberezkin/ajv-errors) lib.

_____

One thing it's better to talk through.

#### How we build schema for different providers?

I see two approaches:
 - One schema for multiple providers. In AJV there are [if/then/else](https://github.com/epoberezkin/ajv/blob/master/KEYWORDS.md#ifthenelse) keywords that I think could be useful here (or maybe there is an easier way?). This approach would make the schema more complicated, though.
- Load provider's schema. This way, `configSchemaHandler.loadSchema(providerName)` should be called after the provider is known, but before plugins are loaded.

____

As always, any feedback is welcome

**Comment by preshetin** [comment]
Looks like this should work for multi-provider case:
```javascript
const awsEventsSchema = require('./awsEventsSchema')
const azureEventsSchema = require('./azureEventsSchema')
const functionNamePattern = "^[a-zA-Z0-9-_]+$";

const schema = {
  type: "object",
  properties: {
    custom: {
      type: "object",
      properties: {}
    },
    provider: {
      type: 'object',
      properties: {
        stage: { type: 'string' },
        variableSyntax: { type: 'string'},
        name: {enum: ['aws', 'azure'] },
        region: {type: 'string'},
        versionFunctions: {type: 'boolean'},
        remoteFunctionData: {},
      }
    },
  },
  allOf: [
    {
      if: { properties: { provider: { properties: { name: {const: 'aws' } } } } },
      then: {
        properties: { functions: getFunctionsSchema(awsEventsSchema) }
      }
    },
    {
      if: { properties: { provider: { properties: { name: {const: 'azure' } } } } },
      then: {
        properties: { functions: getFunctionsSchema(azureEventsSchema) }
      }
    },
  ],
};

function getFunctionsSchema (evSchema) {
  return {
    type: "object",
    patternProperties: {
      [functionNamePattern]: {
        type: "object",
        properties: {
          handler: { type: "string" },
          name: { type: "string" }, // added
          events: evSchema
        },
        required: ["handler"],
        additionalProperties: false
      }
    },
    additionalProperties: false
  }
}
```
----

So if there are no more things to discuss, I'd like to modify the existing PR with changes that we have discussed here

**Comment by medikoo** [comment]
> The solution I'm thinking should only display one message with text like this:
> ```
> "websocketssss" should NOT have additional properties (path: ".functions['someFunc'].events[2]")
> ``` 

Ideally in this case error message I believe should be:

```
Unsupported function event "websocketssss"
```

I think it might be nice to recognize such (few) obvious cases, and handle them specifically so user friendly message is thrown.

e.g. above case can be detected by fact of `'should match some schema in anyOf'` error being thrown on dataPath as `.functions['xxx'].events[x]`.

So I guess whenever validation fails, and we see this error, we should just expose error as `Unsupported function event "x"`

Otherwise for all other (not specifically handled cases) what you propose in `composeErrorMessages` looks great.

> How we build schema for different providers?

Very good question. 

I would conditionally extend schema, with means we already talked about (helpers as `defineFunctionEvent` exposed on `configSchemaHandler`)

e.g in case of AWS, schema that concerns AWS specific `provider` settings, should be added at this logic point: https://github.com/serverless/serverless/blob/08ec261a3cd34e7225f471cbeab8cef605ac61fc/lib/plugins/aws/provider/awsProvider.js#L132
That way it'll be added only if indeed we're in context of AWS service.

Same with schemas for individual events, e.g. in constructor which configures logic that handles `alb` events, schema for this event (via `defineFunctionEvent`) should be added only if we're in context of AWS service (those plugins in general could be optimized to not register hooks, and even require modules when there's no need to, but ofc it's not in scope of this PR).

I think we definitely should not load schemas for two providers in same run (not unless we want to support multi-provider deploys with one shot, and afaik we don't provide such support now). 

So I would not resort to conditional allOf's, that should make things easier.

--- 

I think it'll be good to agree what should be covered in this initial PR.

I propose to provide just validation engine working, with a minimal core config coverage (we probably should cover the structure of the config, but no deep details yet).
 
Having that in, with other PR's we may cover in detail all blanks we leave as e.g. specific schema for each function event. 
It'll be good to open issues with "help-wanted" label for each part that needs to be covered, so we track it, and community is encouraged to help (and point raised here by @DaveLo should rather be covered in PR in which we will define schema for ALB event, and not here in our initial one).

So what should be covered point by point:

1. Core schema (placed in `lib/configSchema`) which covers following:
  - `service` - it covers just two properties, so I believe we can validate it fullly from start
  - `app` - string property validation
  - `org` - string property validation
  - `custom` - object, with free-form structure
   - `plugins` - allow just list of strings, no support for `localPath` and `modules`, we already consider it deprecated, and as it's likely we'll introduce this with v2.0.0, it's ok to be breaking here
  - `resources` - in context of this PR, I think allow object with free-form structure, we will fine tune it dedicated PR
  - `functions` - ideally core should just define and allow `name` and `events` properties.
Other properties should be defined in AWS plugin context, and in context of this PR, I would leave placeholder, which allows additional properties on it, to be overriden by specific schema in other PR. Then similar I would do for each function event, in corresponding plugin configure a schema that allows any configuration for given event (to be fine tuned in dedicated PR's)
  -  `provider` - support just `name`, and in AWS provider plugin allow additional properties (to be replaced with specific schema in dedicated PR)
 - `package` - Possibly validate fully in context of this PR
 - `layers` - Support object map, but with free form content for now (to be fine tuned in dedicated PR)
  - `outputs` - I've just looked into code, and it seems as dead part (not used anywhere). I wouldn't support it. With some other PR I would remove that logic from `service`.

2. Validation should be invoked in `service.validate`. I see that currently `service.validate` has some conditional `stage` validation which jumps in, only if there's `http` event configured.
I think such conditional validation should not be part of our schema. To not introduce regression here I propose to move it to `validate` method of plugin that processes `http` event: https://github.com/serverless/serverless/blob/08ec261a3cd34e7225f471cbeab8cef605ac61fc/lib/plugins/aws/package/compile/events/apiGateway/lib/validate.js#L38

Let me know your thoughts

**Comment by coyoteecd** [comment]
@medikoo I am using [this VSCode plugin](https://github.com/threadheap/serverless-ide-vscode) for validating the serverless.yml file. It gives you autocomplete and validation-on-save within VSCode, including the CloudFormation resources (via official cfn-lint linter). 

Maybe it's an idea that you guys communicate with the author to avoid reinventing some wheels.

**Comment by medikoo** [comment]
@coyoteecd thanks for pointing this. I was not aware of this plugin.

As I see it focuses on CloudFormation validation and in our config picks on places where directly template snippets maybe used. If my assumptions are correct then I think areas of validation we're covering here and one provided by the plugin do not overlap really.

We do not plan to produce JSON schema for CloudFormation parts (it's at stake of validation we run against AWS SDK), also what we're doing here is not really a static analysis. We load all the plugins, resolve variables (some involve remote requests), and then validate the outcome. That'll be difficult to efficiently incorporate into IDE.

**Comment by coyoteecd** [comment]
@medikoo yeah I understand you're doing something different, just wanted to make you aware of what real users of serverless - e.g. me :) - do to address the potential typos in the yml files.

The plugin that I linked [has a schema for the serverless.yml file](https://raw.githubusercontent.com/threadheap/serverless-ide-vscode/master/packages/serverless-framework-schema/schema.json), it's just the github description that focuses on the CloudFormation part; I know because I submitted a couple of pull requests to include all the logs.restApi options. That gives inline autocomplete and documentation on hover so you don't have to refer back to serverless.yml for options. True, it doesn't give validation of values, which is the higher goal you're reaching for.

Maybe what you guys need to do is publish and maintain the serverless.yml schema as part of the serverless project, use it during validation, and make it possible for 3rd parties (such as the plugin in question) to reference it instead of duplicate it. Benefit is when I submit a pull request to serverless that adds a new property, I don't have to submit a matching pull request to the plugin that handles IDE integration.

**Comment by preshetin** [comment]
> Ideally in this case error message I believe should be:
> 
> ```
> Unsupported function event "websocketssss"
> ```

I agree. This message looks more user-friendly.

> I would conditionally extend schema, with means we already talked about (helpers as `defineFunctionEvent` exposed on `configSchemaHandler`)

Thanks for pointing this out. It made me realize, that the Framework has many internal plugins, where the schema can be extended. This way,  the initial `lib/configSchema/index.js` file would be much simpler than I thought. Also, it would not need to use AJV's if/then keywords as I see it now.

However, specifically for the Azure's `http` event, I did not find a plugin where this event is defined (I was curious where in code the schema should be extended). Though this seems out of the scope of this PR.
 
> e.g in case of AWS, schema that concerns AWS specific `provider` settings, should be added at this logic point:
> 
> https://github.com/serverless/serverless/blob/08ec261a3cd34e7225f471cbeab8cef605ac61fc/lib/plugins/aws/provider/awsProvider.js#L132
> 
> 
> That way it'll be added only if indeed we're in context of AWS service.

This also seems a good place to add placeholders for `layers` and `resources` schema parts. Assuming they exist in AWS case only.

> So what should be covered point by point:
> 
> 1. Core schema (placed in `lib/configSchema`) which covers following:

I agree with this list. So am planning to use it as an instruction while implementing the PR.
 
> 1. Validation should be invoked in `service.validate`. I see that currently `service.validate` has some conditional `stage` validation which jumps in, only if there's `http` event configured.
>    I think such conditional validation should not be part of our schema. To not introduce regression here I propose to move it to `validate` method of plugin that processes `http` event: https://github.com/serverless/serverless/blob/08ec261a3cd34e7225f471cbeab8cef605ac61fc/lib/plugins/aws/package/compile/events/apiGateway/lib/validate.js#L38

I'll try to implement this too.

-----

@coyoteecd I've looked through the schema file you provided. Really impressed with the work already made by that plugin authors. If we still go the way we'are discussing here, for individual event schemas I think it would be useful to check how it's done in there.

As for your idea of publishing the schema, I'm hoping @medikoo would share his thoughts on that.

**Comment by medikoo** [comment]
> Maybe what you guys need to do is publish and maintain the serverless.yml schema as part of the serverless project, use it during validation, and make it possible for 3rd parties (such as the plugin in question)

Thanks @coyoteecd for input. That's generally a great idea. Still problem is, that we decided to not prepare any schema here for `serverless.yml`, as is. Due to implications related to variable resolutions and eventual plugin extensions, we decided that this file (as is) should be just validated by YAML parser. 

Schema we prepare is for this config with variables being resolved, also schema we prepare is dynamic in sense, that plugins at initialization are expected to extend this schema (if those extensions are not in place, then config may not validate).

So it's difficult to prepare it for IDE needs, as it'll need to run `serverless` process that resolves variables (in some cases with remote requests) and initiales plugins to ensure schema is complete before validation takes place

**Comment by medikoo** [comment]
> However, specifically for the Azure's http event, I did not find a plugin where this event is defined 

@preshetin If Azure defines `http` event, it'll be in https://github.com/serverless/serverless-azure-functions. Still plugins are out of scope now. Eventually we may later help those most popular to get their schemas right.

It's just AWS provider which has support provided in the core, and it's `http` event support is configured here: https://github.com/serverless/serverless/blob/5017f038d6a8f35fc25ec7a239358a30ca15b745/lib/plugins/aws/package/compile/events/apiGateway/index.js

> This also seems a good place to add placeholders for `layers` and `resources` schema parts. Assuming they exist in AWS case only.

They are defined as provider-agnostic. So I believe basic schema should be in `lib/configSchema`. It's just AWS specific stuff that should be eventually at _provider_ (but I wouldn't that stuff in this PR, eventually just add relaxations, so we do not break)

**Comment by medikoo** [comment]
> The plugin that I linked has a schema for the serverless.yml file, it's

@coyoteecd sorry I didn't look in detail into that when writing first comment.

It indeed looks as very valuable resource, and @preshetin I totally agree, it'll be nice to take advantage what's there.

In terms of publishing the schema, what probably would be nice is to prepare a commands that do following:
- Output generated schema (one including all plugin extensions)
- Output config with resolved variables (one we validate with schema)
- Validate config with schema (where both _config_ and _schema_ can be provided as input).

Having that IDE's will have a room to take advantage of that. 

Still I'd leave it for other PR, as it's a significant task on its own

**Comment by rdsedmundo** [comment]
@medikoo I'd like to see this issue extending (or if not, I could create another) the repository to also provide YAML schema at the root that could be configured in the editor so we'd have autocomplete for the options available. It's only a few seconds, but it can be time-consuming to refer to the documentation every time we forgot about a property.

For example, [this JSON schema for Cypress.io](https://raw.githubusercontent.com/cypress-io/cypress/develop/cli/schema/cypress.schema.json) allows you to configure it inside VSCode and enables autocompletion out of the box.

Reference: https://code.visualstudio.com/docs/languages/json#_json-schemas-and-settings

**Comment by medikoo** [comment]
@rdsedmundo thanks for proposal.

Still due to how Framework works, we focus here on dynamically generated schema, which validates the config after it has all variables resolved, and has all eventual extras as added by the plugins.
Those constraints make it not viable to provide something both static (living with repository) and fully reliable.

Main goal is to quickly pick configuration issues at runtime and report them in meaningful way.

We may of course try to propose a first stage static schema for unresolved config (plain `serverless.yml`). Actually this was the initial idea, still after thinking it through, we abandoned it.
It might be time-taking to achieve, and cannot be complete. We need to support variables being used all over the place (or parts of a schema being held in external files) and we cannot support schema extensions eventually coming from external plugins (which do not live in main repository). At this point I'm not sure if it's worth the effort.

For eventual IDE completion, it might better to resort to CLI commands as I proposed [above](https://github.com/serverless/serverless/issues/6562#issuecomment-591090947)
